### PR TITLE
Converter script

### PR DESCRIPTION
**Pull request by Mara3l**

### CODE REVIEW

File: scripts/SD_converter.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        question, answer = entry.split("Answer:")
+        question = question.strip()
+        answer = answer.strip()
+
+        qa_pair = {"question": question, "answer": answer}
+        jsonl_output.append(json.dumps(qa_pair))
+
+    with open(output_file, "w") as f:
+        for line in jsonl_output:
+            f.write(line + "\n")
```
Comment (hkad98) [inline scripts/SD_converter.py:23]:
```suggestion
            f.write(f"{line}\n")
```

File: scripts/SD_converter.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        if not os.path.isfile(input_file):
+            raise Exception("Input file does not exist!")
+
+        if os.path.isfile(output_file):
+            raise Exception("Output file already exists!")
+
+        if sys.argv[3] == "Jsonl":
+            txt_to_jsonl(input_file=input_file, output_file=output_file)
+        else:
+            raise Exception("Wrong format argument")
```
Comment (hkad98) [inline scripts/SD_converter.py:51]: Use argparse

File: scripts/SD_converter.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        Raises
+        ------
+        Exception
+            If input file doesn't exist, output file already exists, or an invalid format is provided.
+        """
+        if not os.path.isfile(self.input_file):
+            raise Exception("Input file does not exist!")
+
+        if os.path.isfile(self.output_file):
```
Comment (jaceksan) [inline scripts/SD_converter.py:101]: Allow overwriting of existing files. If I change an input file, I want to regenerate the corresponding output file. Force users to delete it manually first is not convenient.

File: scripts/SD_converter.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
+import json
```
Comment (jaceksan) [inline scripts/SD_converter.py:1]: Please, move this file to the main folder, so I do not have to always specify file paths with ../

File: SD_converter.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        qa_pairs = []
+        for entry in entries[1:]:
+            question, answer = entry.split("Answer:")
+            question = question.strip()
+            answer = answer.strip()
+            qa_pairs.append((answer, question))
+
+        return qa_pairs
+
+    def _to_jsonl(self, qa_pairs: List[Tuple[str, str]], output_file: str) -> None:
```
Comment (hkad98) [inline SD_converter.py:36]: Name it according to format ;)

File: SD_converter.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        jsonl_output = []
+        for answer, question in qa_pairs:
+            qa_pair = {"question": question, "answer": answer}
+            jsonl_output.append(json.dumps(qa_pair))
+
+        with open(output_file, "w") as f:
+            for line in jsonl_output:
+                f.write(f"{line}\n")
+
+    def _to_choices(self, qa_pairs: List[Tuple[str, str]], output_file: str) -> None:
```
Comment (hkad98) [inline SD_converter.py:54]: Name it according to format ;)

File: SD_converter.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        Converts a list of Question-Answer pairs to the Choices format and writes to the output file.
+
+        Args:
+            qa_pairs (List[Tuple[str, str]]): A list of Question-Answer pairs.
+            output_file (str): Path to the output file.
+        """
+        choices = []
+        for answer, question in qa_pairs:
+            # Choices is a list, hence the use of square brackets ([]).
+            qa_pair = {"Question": question, "choices": "[" + answer + "]"}
```
Comment (hkad98) [inline SD_converter.py:65]: See format one more time. "choices" should be a list of string not a string.

File: SD_converter.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            choices.append(json.dumps(qa_pair))
+
+        with open(output_file, "w") as f:
+            for line in choices:
+                f.write(f"{line}\n")
```
Comment (hkad98) [inline SD_converter.py:70]: Repeating code. Make it a separate function.

File: SD_converter.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            print(f"Output file: {output_file} exists, overwriting.")
+
+        qa_pairs = self._parse_input_file(input_file)
+
+        if format == "Jsonl":
+            self._to_jsonl(qa_pairs=qa_pairs, output_file=output_file)
+        elif format == "choices":
+            self._to_choices(qa_pairs=qa_pairs, output_file=output_file)
+        else:
+            raise Exception("Wrong format argument")
```
Comment (hkad98) [inline SD_converter.py:94]: Wrong formats. Make it a match - case, so we can extend it in the future.

File: SD_converter.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    """The main entry point of the script. Runs when the script is called from the command line."""
+    import argparse
+
+    parser = argparse.ArgumentParser(
+        description="This is a helper script to convert between different AI fine-tuning notations."
+    )
+    parser.add_argument("-i", "--input", type=str, help="Path to input file")
+    parser.add_argument("-o", "--output", type=str, help="Path to output file", default="output.jsonl")
+    parser.add_argument(
+        "-f", "--format", type=str, choices=["oAI", "HF"], default="oAI", help="Output format (Jsonl or choices)"
```
Comment (hkad98) [inline SD_converter.py:107]: Help is outdated. I suggested choices to be ["oai", "hf"]

File: SD_converter.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
+import json
+import os
+from typing import List, Tuple
+
+
+class Converter:
+    """A class to represent a converter for AI fine-tuning notations."""
+
+    def __init__(self):
+        return
```
Comment (hkad98) [inline SD_converter.py:10]: This should not be necessary.

File: SD_converter.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        Args:
+            qa_pairs (List[Tuple[str, str]]): A list of Question-Answer pairs.
+
+        Returns:
+            List[str]: Content of the final .jsonl file
+        """
+
+        content = []
+        for answer, question in qa_pairs:
+            qa_pair = {"question": question, "answer": answer}
```
Comment (hkad98) [inline SD_converter.py:73]: Not correct. See [openai documentation](https://platform.openai.com/docs/guides/fine-tuning/prepare-training-data).

`{"prompt": "<prompt text>", "completion": "<ideal generated text>"}`

File: SD_converter.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        Args:
+            qa_pairs (List[Tuple[str, str]]): A list of Question-Answer pairs.
+
+        Returns:
+            List[str]: Content of the final .jsonl file
+        """
+        content = []
+        for answer, question in qa_pairs:
+            # Choices is a list, hence the use of square brackets ([]).
+            qa_pair = {"Question": question, "choices": [answer]}
```
Comment (hkad98) [inline SD_converter.py:90]: This seems to be invalid as well. See [example](https://github.com/mosaicml/llm-foundry/blob/688cc52668e9eb514a8a6b3dc14288cd9babe8af/scripts/train/finetune_example/train.jsonl#L4).

The keys are **choices** it is correct, but then there is **query**.

Your code will always have cardinality 1:1. For each query there will be one choice. Is it intentional?

File: SD_converter.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+        qa_pairs = self._parse_input_file(input_file)
+
+        match format:
+            case "oai":
+                self._to_file(output_file=output_file, content=self._to_oai(qa_pairs=qa_pairs))
+            case "hf":
+                self._to_file(output_file=output_file, content=self._to_hf(qa_pairs=qa_pairs))
+            case _:
+                raise Exception("Wrong format argument")
```
Comment (hkad98) [inline SD_converter.py:117]:
```suggestion
        match format:
            case "oai":
                content = self._to_oai(qa_pairs=qa_pairs)             
            case "hf":
                content = self._to_hf(qa_pairs=qa_pairs)     
            case _:
                raise Exception("Wrong format argument")
        self._to_file(output_file=output_file, content=content)      
```

File: source_documents/twerfon.txt [hunk 15]
```diff
@@ -1,3 +1,5 @@
+Question: What is the twerfon?
```
Comment (hkad98) [inline source_documents/twerfon.txt:1]: Probably missing new line after `Question:`

File: source_documents/example_ldm_demo.txt [hunk 16]
```diff
@@ -69,4 +72,4 @@ Dataset with ID "products" has the following attributes:
 - Attribute with ID "products.category" has title "Category"
 
 Question: what is the title of the dataset with ID "campaign_channels"?
-Answer: the title is "Campaign channels"
\ No newline at end of file
+Answer: the title is "Campaign channels"
```
Comment (hkad98) [inline source_documents/example_ldm_demo.txt:75]: Missing new line after `Answer:`

File: source_documents/example_ldm_33f6ff605e0544ef8042c976bc103b0c.txt [hunk 17]
```diff
@@ -112,4 +114,4 @@ Dataset with ID "supplier" has the following attributes:
 - Attribute with ID "s_suppkey" has title "S suppkey"
 
 Question: what is the title of the dataset with ID "customer"?
-Answer: the title is "Customer"
\ No newline at end of file
+Answer: the title is "Customer"
```
Comment (hkad98) [inline source_documents/example_ldm_33f6ff605e0544ef8042c976bc103b0c.txt:117]: Missing new line after Answer:

File: SD_converter.py (hunk 13 above)
Comment (Mara3l) [reply SD_converter.py:90]: The cardinality is intentional.

**Pull request merged**